File: PRNumberBiz/app/src/main/java/com/pplus/prnumberbiz/apps/setting/ui/AlarmContainerActivity.java
package com.pplus.prnumberbiz.apps.setting.ui;

import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.widget.CompoundButton;

import com.pple.pplus.utils.part.utils.StringUtils;
import com.pplus.prnumberbiz.Const;
import com.pplus.prnumberbiz.R;
import com.pplus.prnumberbiz.apps.bol.ui.PointConfigActivity;
import com.pplus.prnumberbiz.apps.cash.ui.CashConfigActivity2;
import com.pplus.prnumberbiz.apps.common.builder.AlertBuilder;
import com.pplus.prnumberbiz.apps.common.builder.OnAlertResultListener;
import com.pplus.prnumberbiz.apps.common.builder.data.AlertData;
import com.pplus.prnumberbiz.apps.common.mgmt.LoginInfoManager;
import com.pplus.prnumberbiz.apps.common.toolbar.ImplToolbar;
import com.pplus.prnumberbiz.apps.common.toolbar.OnToolbarListener;
import com.pplus.prnumberbiz.apps.common.toolbar.ToolbarOption;
import com.pplus.prnumberbiz.apps.common.ui.base.BaseActivity;
import com.pplus.prnumberbiz.apps.common.ui.custom.SafeSwitchCompat;
import com.pplus.prnumberbiz.apps.main.ui.BizMainActivity;
import com.pplus.prnumberbiz.apps.post.ui.PostDetailActivity;
import com.pplus.prnumberbiz.apps.setting.data.AlarmAdapter;
import com.pplus.prnumberbiz.core.code.common.MoveType1Code;
import com.pplus.prnumberbiz.core.code.common.MoveType2Code;
import com.pplus.prnumberbiz.core.network.ApiBuilder;
import com.pplus.prnumberbiz.core.network.model.dto.Bol;
import com.pplus.prnumberbiz.core.network.model.dto.Msg;
import com.pplus.prnumberbiz.core.network.model.dto.Notice;
import com.pplus.prnumberbiz.core.network.model.dto.Post;
import com.pplus.prnumberbiz.core.network.model.response.NewResultResponse;
import com.pplus.prnumberbiz.core.util.PplusCommonUtil;

import java.util.HashMap;
import java.util.Map;

import network.common.PplusCallback;
import retrofit2.Call;

public class AlarmContainerActivity extends BaseActivity implements ImplToolbar{

    private LinearLayoutManager mLayoutManager;
    private AlarmAdapter mAdapter;

    private int mPage = 1, mTotalCount = 0;
    private boolean mLockListView = true;

    @Override
    public String getPID(){

        return "";
    }

    @Override
    public int getLayoutRes(){

        return R.layout.activity_alarm_container;
    }

    @Override
    public void initializeView(Bundle savedInstanceState){

        RecyclerView recyclerView = (RecyclerView)findViewById(R.id.recycler_alarm);
        mLayoutManager = new LinearLayoutManager(this);
        mLayoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        recyclerView.setLayoutManager(mLayoutManager);
        mAdapter = new AlarmAdapter(this);
        recyclerView.setAdapter(mAdapter);
        mAdapter.setActivity(this);

        mAdapter.setOnItemClickListener(new AlarmAdapter.OnItemClickListener(){

            @Override
            public void onItemClick(int position){


                final Msg item = mAdapter.getItem(position);

                MoveType1Code moveType1Code = null;
                MoveType2Code moveType2Code = null;
                if(item.getMoveType2() != null) {
                    try {
                        moveType1Code = MoveType1Code.valueOf(item.getMoveType1());
                        moveType2Code = MoveType2Code.valueOf(item.getMoveType2());
                    } catch (Exception e) {

                    }
                }

                if(moveType1Code == null){
                    return;
                }

                if(moveType1Code == MoveType1Code.inner){
                    // 스키마 호출
                    if(moveType2Code == null){
                        return;
                    }
                    // 스키마 호출
                    switch (moveType2Code) {
                        case pageDetail:
                            Intent intent = new Intent(AlarmContainerActivity.this, BizMainActivity.class);
                            //                        Page page = new Page();
                            //                        page.setNo(Long.valueOf(item.getMoveTarget().getNo()));
                            //                        intent.putExtra(Const.PAGE, page);
                            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
                            startActivity(intent);
                            callRead(""+item.getNo());
                            break;
                        case postDetail:
                            intent = new Intent(AlarmContainerActivity.this, PostDetailActivity.class);
                            Post post = new Post();
                            post.setNo(item.getMoveTarget().getNo());
                            intent.putExtra(Const.DATA, post);
                            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
                            startActivity(intent);
                            callRead(""+item.getNo());
                            break;
//                        case noteDetail:
//                            intent = new Intent(AlarmContainerActivity.this, NoteDetailActivity.class);
//                            intent.putExtra(Const.NO, item.getMoveTarget().getNo());
//                            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
//                            startActivity(intent);
//                            callRead(""+item.getNo());
//                            break;
                        case plus:
//                            intent = new Intent(AlarmContainerActivity.this, PlusActivity.class);
//                            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
//                            startActivity(intent);
                            break;
                        case bolHistory:
                            intent = new Intent(AlarmContainerActivity.this, PointConfigActivity.class);
                            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
                            startActivity(intent);
                            break;
                        case cashHistory:
                            intent = new Intent(AlarmContainerActivity.this, CashConfigActivity2.class);
                            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
                            startActivity(intent);
                            break;
                        case msgbox:
                            break;
                        case offer:
                            break;
//                        case adCoupon:
//                            intent = new Intent(AlarmContainerActivity.this, RealTimeCouponDetailActivity.class);
//                            Advertise advertise = new Advertise();
//                            advertise.setNo(item.getNo());
//                            intent.putExtra(Const.ADVERTISE, advertise);
//                            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP );
//                            startActivity(intent);
//                            break;
//                        case adPost:
//                            intent = new Intent(AlarmContainerActivity.this, RealTimePostDetailActivity.class);
//                            advertise = new Advertise();
//                            advertise.setNo(item.getNo());
//                            intent.putExtra(Const.ADVERTISE, advertise);
//                            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP );
//                            startActivity(intent);
//                            break;
                        case eventWin:
                            break;
                        case eventDetail:
                            break;
                        case noticeDetail:
                            intent = new Intent(AlarmContainerActivity.this, NoticeDetailActivity.class);
                            Notice notice = new Notice();
                            notice.setNo(item.getNo());
                            intent.putExtra(Const.NOTICE, notice);
                            startActivity(intent);
                            break;
                        case bolDetail:
                            intent = new Intent(AlarmContainerActivity.this, PointConfigActivity.class);
                            Bol bol = new Bol();
                            bol.setNo(item.getNo());
                            intent.putExtra(Const.DATA, bol);
                            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
                            startActivity(intent);
                            break;
                    }
                }else{
                    if(StringUtils.isNotEmpty(item.getMoveTargetString())){
                        PplusCommonUtil.Companion.openChromeWebView(AlarmContainerActivity.this, item.getMoveTargetString());
                    }
                }
            }
        });

        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener(){

            int pastVisibleItems, visibleItemCount, totalItemCount;

            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy){

                super.onScrolled(recyclerView, dx, dy);
                visibleItemCount = mLayoutManager.getChildCount();
                totalItemCount = mLayoutManager.getItemCount();
                pastVisibleItems = mLayoutManager.findFirstVisibleItemPosition();
                if(!mLockListView) {
                    if((totalItemCount < mTotalCount) && (visibleItemCount + pastVisibleItems) >= (totalItemCount * Const.NEXT_RECYCLERVIEW_MARGIN)) {
                        mPage++;
                        listCall(mPage);
                    }
                }
            }
        });
        getCount();
    }

    private void callRead(String no){
        Map<String, String> params = new HashMap<>();
        params.put("no", no);
        ApiBuilder.create().readComplete(params).setCallback(new PplusCallback<NewResultResponse<Object>>(){

            @Override
            public void onResponse(Call<NewResultResponse<Object>> call, NewResultResponse<Object> response){

            }

            @Override
            public void onFailure(Call<NewResultResponse<Object>> call, Throwable t, NewResultResponse<Object> response){

            }
        }).build().call();
    }

    private void getCount(){

        ApiBuilder.create().getMsgCountInBox().setCallback(new PplusCallback<NewResultResponse<Integer>>(){

            @Override
            public void onResponse(Call<NewResultResponse<Integer>> call, NewResultResponse<Integer> response){
                mTotalCount = response.getData();
                mPage = 1;
                mAdapter.clear();
                listCall(mPage);
            }

            @Override
            public void onFailure(Call<NewResultResponse<Integer>> call, Throwable t, NewResultResponse<Integer> response){

            }
        }).build().call();
    }

    private void listCall(int page){
        Map<String, String> params = new HashMap<>();
        params.put("pg", ""+page);
        showProgress("");
        ApiBuilder.create().getMsgListInBox(params).setCallback(new PplusCallback<NewResultResponse<Msg>>(){

            @Override
            public void onResponse(Call<NewResultResponse<Msg>> call, NewResultResponse<Msg> response){
                hideProgress();
                mAdapter.addAll(response.getDatas());
            }

            @Override
            public void onFailure(Call<NewResultResponse<Msg>> call, Throwable t, NewResultResponse<Msg> response){
                hideProgress();
            }
        }).build().call();
    }

    public void delete(Long no){

        Map<String, String> params = new HashMap<>();
        params.put("no", ""+no);
        showProgress("");
        ApiBuilder.create().deleteMsgInBox(params).setCallback(new PplusCallback<NewResultResponse<Object>>(){

            @Override
            public void onResponse(Call<NewResultResponse<Object>> call, NewResultResponse<Object> response){
                hideProgress();
                showAlert(R.string.msg_delete_complete);
                getCount();
            }

            @Override
            public void onFailure(Call<NewResultResponse<Object>> call, Throwable t, NewResultResponse<Object> response){
                hideProgress();
            }
        }).build().call();
    }


    @NonNull
    @Override
    public ToolbarOption getToolbarOption(){

        ToolbarOption toolbarOption = new ToolbarOption(this);
        toolbarOption.initializeDefaultToolbar(getString(R.string.word_alarm_container), ToolbarOption.ToolbarMenu.LEFT);
        View view = getLayoutInflater().inflate(R.layout.item_top_switch, null);
        final SafeSwitchCompat switchCompat = (SafeSwitchCompat) view.findViewById(R.id.switch_top_right);
        if(LoginInfoManager.getInstance().getUser().getDevice().getInstalledApp() != null){
            switchCompat.setSafeCheck(LoginInfoManager.getInstance().getUser().getDevice().getInstalledApp().isPushActivate(), SafeSwitchCompat.IGNORE);
        }else{
            switchCompat.setSafeCheck(false, SafeSwitchCompat.IGNORE);
        }

        switchCompat.setOnSafeCheckedListener(new SafeSwitchCompat.OnSafeCheckedListener(){

            @Override
            public void onAlwaysCalledListener(CompoundButton buttonView, boolean isChecked){

            }

            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked){

                if(isChecked) {
                    String pushMask = "1111111111111111";
                    pushUpdate(pushMask, true);
                } else {
                    AlertBuilder.Builder builder = new AlertBuilder.Builder();
                    builder.setTitle(getString(R.string.word_notice_alert));
                    builder.addContents(new AlertData.MessageData(getString(R.string.msg_alert_alarm_off), AlertBuilder.MESSAGE_TYPE.TEXT, 3));
                    builder.setLeftText(getString(R.string.word_cancel)).setRightText(getString(R.string.word_confirm));
                    builder.setOnAlertResultListener(new OnAlertResultListener(){

                        @Override
                        public void onCancel(){
                            switchCompat.setSafeCheck(true, SafeSwitchCompat.IGNORE);
                        }

                        @Override
                        public void onResult(AlertBuilder.EVENT_ALERT event_alert){

                            switch (event_alert) {
                                case RIGHT:
                                    String pushMask = "0000000000000000";
                                    pushUpdate(pushMask, false);
                                    break;
                                case LEFT:
                                    switchCompat.setSafeCheck(true, SafeSwitchCompat.IGNORE);
                                    break;
                            }

                        }
                    }).builder().show(AlarmContainerActivity.this);
                }
            }
        });
        toolbarOption.setToolbarMenu(ToolbarOption.ToolbarMenu.RIGHT, view, 0);
        return toolbarOption;
    }

    private void pushUpdate(final String pushMask, final boolean pushActivate){

        Map<String, String> params = new HashMap<>();
        params.put("pushActivate", String.valueOf(pushActivate));
        params.put("pushMask", pushMask);

        showProgress("");
        ApiBuilder.create().updatePushConfig(params).setCallback(new PplusCallback<NewResultResponse<Object>>(){

            @Override
            public void onResponse(Call<NewResultResponse<Object>> call, NewResultResponse<Object> response){

                hideProgress();
                LoginInfoManager.getInstance().getUser().getDevice().getInstalledApp().setPushActivate(pushActivate);
                LoginInfoManager.getInstance().getUser().getDevice().getInstalledApp().setPushMask(pushMask);
            }

            @Override
            public void onFailure(Call<NewResultResponse<Object>> call, Throwable t, NewResultResponse<Object> response){

                hideProgress();
            }
        }).build().call();
    }

    @Override
    public OnToolbarListener getOnToolbarClickListener(){

        return new OnToolbarListener(){

            @Override
            public void onClick(View v, ToolbarOption.ToolbarMenu toolbarMenu, Object tag){
                switch (toolbarMenu) {
                    case LEFT:
                        if(tag.equals(1)) {
                            onBackPressed();
                        }
                        break;
                }
            }
        };
    }
}

File: PRNumber/app/src/main/java/com/pplus/prnumberuser/apps/setting/ui/AlarmContainerActivity.kt
package com.pplus.prnumberuser.apps.setting.ui

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.CompoundButton
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.pplus.networks.common.PplusCallback
import com.pplus.prnumberuser.Const
import com.pplus.prnumberuser.R
import com.pplus.prnumberuser.apps.bol.ui.BolConfigActivity
import com.pplus.prnumberuser.apps.common.builder.AlertBuilder
import com.pplus.prnumberuser.apps.common.builder.OnAlertResultListener
import com.pplus.prnumberuser.apps.common.builder.data.AlertData
import com.pplus.prnumberuser.apps.common.mgmt.DeliveryAddressManager
import com.pplus.prnumberuser.apps.common.mgmt.LoginInfoManager
import com.pplus.prnumberuser.apps.common.toolbar.ImplToolbar
import com.pplus.prnumberuser.apps.common.toolbar.OnToolbarListener
import com.pplus.prnumberuser.apps.common.toolbar.ToolbarOption
import com.pplus.prnumberuser.apps.common.ui.base.BaseActivity
import com.pplus.prnumberuser.apps.common.ui.custom.SafeSwitchCompat
import com.pplus.prnumberuser.apps.delivery.ui.DeliveryAddressSetActivity
import com.pplus.prnumberuser.apps.event.ui.EventDetailActivity
import com.pplus.prnumberuser.apps.event.ui.EventImpressionActivity
import com.pplus.prnumberuser.apps.event.ui.EventMoveDetailActivity
import com.pplus.prnumberuser.apps.menu.ui.OrderPurchaseHistoryDetailActivity
import com.pplus.prnumberuser.apps.menu.ui.TicketPurchaseHistoryDetailActivity
import com.pplus.prnumberuser.apps.page.ui.PageDeliveryMenuActivity
import com.pplus.prnumberuser.apps.product.ui.ProductShipDetailActivity
import com.pplus.prnumberuser.apps.product.ui.PurchaseHistoryActivity
import com.pplus.prnumberuser.apps.setting.data.NotificationBoxAdapter
import com.pplus.prnumberuser.core.code.common.EventType
import com.pplus.prnumberuser.core.code.common.MoveType2Code
import com.pplus.prnumberuser.core.network.ApiBuilder
import com.pplus.prnumberuser.core.network.model.dto.*
import com.pplus.prnumberuser.core.network.model.response.NewResultResponse
import com.pplus.prnumberuser.core.network.model.response.SubResultResponse
import com.pplus.prnumberuser.core.util.PplusCommonUtil
import com.pplus.prnumberuser.databinding.ActivityAlarmContainerBinding
import com.pplus.prnumberuser.databinding.ItemTopSwitchBinding
import com.pplus.utils.part.utils.StringUtils
import retrofit2.Call
import java.util.*

class AlarmContainerActivity : BaseActivity(), ImplToolbar {

    private var mLayoutManager: LinearLayoutManager? = null
    private var mAdapter: NotificationBoxAdapter? = null

    private var mTotalCount: Int = 0
    private var mLockListView = false
    private var mPaging = 1
    private var mIsLast = false

    override fun getPID(): String? {
        return ""
    }

    private lateinit var binding: ActivityAlarmContainerBinding

    override fun getLayoutView(): View {
        binding = ActivityAlarmContainerBinding.inflate(layoutInflater)
        return binding.root
    }

    override fun initializeView(savedInstanceState: Bundle?) {

        mLayoutManager = LinearLayoutManager(this)
        mLayoutManager!!.orientation = LinearLayoutManager.VERTICAL
        binding.recyclerAlarm.layoutManager = mLayoutManager
        mAdapter = NotificationBoxAdapter()
        binding.recyclerAlarm.adapter = mAdapter

        mAdapter!!.listener = object : NotificationBoxAdapter.OnItemClickListener {
            override fun onItemClick(position: Int) {
                val item = mAdapter!!.getItem(position)


                if (item.moveType1 == "inner") {
                    if (item.moveType2 == null) {
                        return
                    }
                    val intent: Intent // 스키마 호출
                    when (item.moveType2) {
                        MoveType2Code.main.name -> {
                        }
                        MoveType2Code.pageDetail.name -> {

                            val page = Page()
                            page.no = item.moveSeqNo
                            PplusCommonUtil.goPage(this@AlarmContainerActivity, page, 0, 0)
                        }
                        MoveType2Code.productDetail.name -> {
                            intent = Intent(this@AlarmContainerActivity, ProductShipDetailActivity::class.java)
                            val productPrice = ProductPrice()
                            productPrice.seqNo = item.moveSeqNo
                            intent.putExtra(Const.DATA, productPrice)
                            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                            startActivity(intent)
                        }
                        MoveType2Code.buyCancel.name -> {
                            intent = Intent(this@AlarmContainerActivity, PurchaseHistoryActivity::class.java)
                            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                            startActivity(intent)
                        }
                        MoveType2Code.bolHistory.name -> {
                            intent = Intent(this@AlarmContainerActivity, BolConfigActivity::class.java)
                            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                            startActivity(intent)
                        }
                        MoveType2Code.eventWin.name -> {
                            intent = Intent(this@AlarmContainerActivity, EventImpressionActivity::class.java)
                            val event = Event()
                            event.no = item.moveSeqNo
                            intent.putExtra(Const.DATA, event)
                            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                            startActivity(intent)
                        }
                        MoveType2Code.eventDetail.name -> {

                            val params = HashMap<String, String>()
                            params["no"] = "" + item.moveSeqNo

                            ApiBuilder.create().getEvent(params).setCallback(object : PplusCallback<NewResultResponse<Event>> {

                                override fun onResponse(call: Call<NewResultResponse<Event>>, response: NewResultResponse<Event>) {

                                    val event = response.data
                                    if (event != null) {
                                        if (StringUtils.isNotEmpty(event.primaryType) && event.primaryType == EventType.PrimaryType.move.name) {
                                            val intent = Intent(this@AlarmContainerActivity, EventMoveDetailActivity::class.java)
                                            intent.putExtra(Const.DATA, event)
                                            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                                            startActivity(intent)
                                        } else {
                                            val intent = Intent(this@AlarmContainerActivity, EventDetailActivity::class.java)
                                            intent.putExtra(Const.DATA, event)
                                            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                                            startActivity(intent)
                                        }
                                    }
                                }

                                override fun onFailure(call: Call<NewResultResponse<Event>>, t: Throwable, response: NewResultResponse<Event>) {

                                }
                            }).build().call()
                        }
                        MoveType2Code.noticeDetail.name -> {
                            intent = Intent(this@AlarmContainerActivity, NoticeDetailActivity::class.java)
                            val notice = Notice()
                            notice.no = item.moveSeqNo
                            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                            intent.putExtra(Const.NOTICE, notice)
                            startActivity(intent)
                        }

                        MoveType2Code.bolDetail.name -> {
                            intent = Intent(this@AlarmContainerActivity, BolConfigActivity::class.java)
                            val bol = Bol()
                            bol.no = item.moveSeqNo
                            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                            intent.putExtra(Const.DATA, bol)
                            startActivity(intent)
                        }
                        MoveType2Code.orderDetail.name -> {
                            if (LoginInfoManager.getInstance().isMember) {
                                val intent = Intent(this@AlarmContainerActivity, OrderPurchaseHistoryDetailActivity::class.java)
                                val orderPurchase = OrderPurchase()
                                orderPurchase.seqNo = item.moveSeqNo
                                intent.putExtra(Const.DATA, orderPurchase)
                                startActivity(intent)
                            }
                        }
                        MoveType2Code.ticketDetail.name -> {
                            if (LoginInfoManager.getInstance().isMember) {
                                val intent = Intent(this@AlarmContainerActivity, TicketPurchaseHistoryDetailActivity::class.java)
                                val orderPurchase = OrderPurchase()
                                orderPurchase.seqNo = item.moveSeqNo
                                intent.putExtra(Const.DATA, orderPurchase)
                                startActivity(intent)
                            }
                        }
                        MoveType2Code.menu.name -> {
                            val deliveryAddressList = DeliveryAddressManager.getInstance().deliveryAddressList
                            if(deliveryAddressList != null && deliveryAddressList.isNotEmpty()){
                                val intent = Intent(this@AlarmContainerActivity, PageDeliveryMenuActivity::class.java)
                                val page = Page2()
                                page.seqNo = item.moveSeqNo
                                intent.putExtra(Const.DATA, page)
                                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                                startActivity(intent)
                            }else{
                                val intent = Intent(this@AlarmContainerActivity, DeliveryAddressSetActivity::class.java)
                                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                                startActivity(intent)
                            }
                        }
                    }
                } else {
                    if (StringUtils.isNotEmpty(item.moveString)) {
                        PplusCommonUtil.openChromeWebView(this@AlarmContainerActivity, item.moveString!!)
                    }
                }
            }
        }

        binding.recyclerAlarm.addOnScrollListener(object : RecyclerView.OnScrollListener() {

            internal var pastVisibleItems: Int = 0
            internal var visibleItemCount: Int = 0
            internal var totalItemCount: Int = 0

            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {

                super.onScrolled(recyclerView, dx, dy)

                visibleItemCount = mLayoutManager!!.childCount
                totalItemCount = mLayoutManager!!.itemCount
                pastVisibleItems = mLayoutManager!!.findFirstVisibleItemPosition()
                if (!mLockListView) {
                    if (!mIsLast && visibleItemCount + pastVisibleItems >= totalItemCount * Const.NEXT_RECYCLERVIEW_MARGIN) {
                        mPaging++
                        listCall(mPaging)
                    }
                }
            }
        })

        mPaging = 0
        listCall(mPaging)
    }

    private fun listCall(page: Int) {

        mLockListView = true
        val params = HashMap<String, String>()
        params["page"] = page.toString()
        showProgress("")
        mLockListView = true
        ApiBuilder.create().getNotificationBoxList(params).setCallback(object : PplusCallback<NewResultResponse<SubResultResponse<NotificationBox>>> {

            override fun onResponse(call: Call<NewResultResponse<SubResultResponse<NotificationBox>>>?, response: NewResultResponse<SubResultResponse<NotificationBox>>?) {
                mLockListView = false
                hideProgress()
                if (response?.data != null) {

                    mIsLast = response.data.last!!
                    if (response.data.first!!) {
                        mTotalCount = response.data.totalElements!!
                        mAdapter!!.clear()
                        if (mTotalCount == 0) {
                            binding.textAlarmContainerNotExist.visibility = View.VISIBLE
                        } else {
                            binding.textAlarmContainerNotExist.visibility = View.GONE
                        }
                    }

                    mLockListView = false
                    mAdapter!!.addAll(response.data.content!!)
                }
            }

            override fun onFailure(call: Call<NewResultResponse<SubResultResponse<NotificationBox>>>?, t: Throwable?, response: NewResultResponse<SubResultResponse<NotificationBox>>?) {
                mLockListView = false
                hideProgress()
            }
        }).build().call()
    }

    fun delete(seqNo: Long) {

        val params = HashMap<String, String>()
        params["seqNo"] = seqNo.toString()
        showProgress("")
        ApiBuilder.create().notificationBoxDelete(params).setCallback(object : PplusCallback<NewResultResponse<Any>> {

            override fun onResponse(call: Call<NewResultResponse<Any>>, response: NewResultResponse<Any>) {

                hideProgress()
                showAlert(R.string.msg_delete_complete)
                mPaging = 0
                listCall(mPaging)
            }

            override fun onFailure(call: Call<NewResultResponse<Any>>, t: Throwable, response: NewResultResponse<Any>) {

                hideProgress()
            }
        }).build().call()
    }


    override fun getToolbarOption(): ToolbarOption {

        val toolbarOption = ToolbarOption(this)
        toolbarOption.initializeDefaultToolbar(getString(R.string.word_alarm_container), ToolbarOption.ToolbarMenu.LEFT)
        val topSwitchBinding = ItemTopSwitchBinding.inflate(layoutInflater)
        if (LoginInfoManager.getInstance().user.device!!.installedApp != null) {
            topSwitchBinding.switchTopRight.setSafeCheck(LoginInfoManager.getInstance().user.device!!.installedApp.isPushActivate, SafeSwitchCompat.IGNORE)
        } else {
            topSwitchBinding.switchTopRight.setSafeCheck(false, SafeSwitchCompat.IGNORE)
        }

        topSwitchBinding.switchTopRight.onSafeCheckedListener = object : SafeSwitchCompat.OnSafeCheckedListener {

            override fun onAlwaysCalledListener(buttonView: CompoundButton, isChecked: Boolean) {

            }

            override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {

                if (isChecked) {
                    val pushMask = "1111111111111111"
                    pushUpdate(pushMask, true)
                } else {
                    val builder = AlertBuilder.Builder()
                    builder.setTitle(getString(R.string.word_notice_alert))
                    builder.addContents(AlertData.MessageData(getString(R.string.msg_alert_alarm_off), AlertBuilder.MESSAGE_TYPE.TEXT, 3))
                    builder.setLeftText(getString(R.string.word_cancel)).setRightText(getString(R.string.word_confirm))
                    builder.setOnAlertResultListener(object : OnAlertResultListener {

                        override fun onCancel() {
                            topSwitchBinding.switchTopRight.setSafeCheck(true, SafeSwitchCompat.IGNORE)
                        }

                        override fun onResult(event_alert: AlertBuilder.EVENT_ALERT) {

                            when (event_alert) {
                                AlertBuilder.EVENT_ALERT.RIGHT -> {
                                    val pushMask = "0000000000000000"
                                    pushUpdate(pushMask, false)
                                }
                                AlertBuilder.EVENT_ALERT.LEFT -> topSwitchBinding.switchTopRight.setSafeCheck(true, SafeSwitchCompat.IGNORE)
                            }

                        }
                    }).builder().show(this@AlarmContainerActivity)
                }
            }
        }
        toolbarOption.setToolbarMenu(ToolbarOption.ToolbarMenu.RIGHT, topSwitchBinding.root, 0)
        return toolbarOption
    }

    private fun pushUpdate(pushMask: String, pushActivate: Boolean) {

        val params = HashMap<String, String>()
        params["pushActivate"] = pushActivate.toString()
        params["pushMask"] = pushMask

        showProgress("")
        ApiBuilder.create().updatePushConfig(params).setCallback(object : PplusCallback<NewResultResponse<Any>> {

            override fun onResponse(call: Call<NewResultResponse<Any>>, response: NewResultResponse<Any>) {

                hideProgress()
                LoginInfoManager.getInstance().user.device!!.installedApp.isPushActivate = pushActivate
                LoginInfoManager.getInstance().user.device!!.installedApp.pushMask = pushMask
            }

            override fun onFailure(call: Call<NewResultResponse<Any>>, t: Throwable, response: NewResultResponse<Any>) {

                hideProgress()
            }
        }).build().call()
    }

    override fun getOnToolbarClickListener(): OnToolbarListener {

        return object : OnToolbarListener {
            override fun onClick(v: View?, toolbarMenu: ToolbarOption.ToolbarMenu?, tag: Any?) {
                when (toolbarMenu) {
                    ToolbarOption.ToolbarMenu.LEFT -> if (tag == 1) {
                        onBackPressed()
                    }
                }
            }
        }
    }
}
